compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -375,6 +424,10 @@ fn push_debuginfo_type_name<'tcx>(
         }
     }
 
+    if type_name_cache.borrow_mut().insert((&t, qualified), output.clone()).is_some() {
+        bug!("type name is already in the type name cache!");
+    }
```

Comment:
I think this is in the wrong place. `push_debuginfo_typename()` can be called recursively, so adding the contents here might insert half-built names into the cache. It looks like a likely cause for the test failures we are seeing. Try to move this to the end of `compute_debuginfo_type_name()` where we know that we are always dealing with complete type names.